app/src/main/java/com/nextcloud/talk/shareditems/activities/SharedItemsActivity.kt

```diff
@@ -55,36 +59,31 @@ class SharedItemsActivity : AppCompatActivity() {
         supportActionBar?.title = conversationName
         supportActionBar?.setDisplayHomeAsUpEnabled(true)
 
-        viewModel = ViewModelProvider(
-            this,
-            SharedItemsViewModel.Factory(userEntity, roomToken, currentTab)
-        ).get(SharedItemsViewModel::class.java)
+        viewModel = ViewModelProvider(this, viewModelFactory)[SharedItemsViewModel::class.java]
+        viewModel.initialize(userEntity, roomToken, SharedItemType.MEDIA)
 
-        viewModel.sharedItemType.observe(this) {
+        viewModel.sharedItemTypes.observe(this) {
             initTabs(it)
         }
 
         viewModel.sharedItems.observe(this) {
             Log.d(TAG, "Items received: $it")
 
-            if (currentTab == SharedItemType.MEDIA) {
-                val adapter = SharedItemsGridAdapter()
-                adapter.items = it.items
-                adapter.authHeader = it.authHeader
-                binding.imageRecycler.adapter = adapter
-
+            val pair = if (viewModel.currentItemType == SharedItemType.MEDIA) {
+                val adapter = SharedItemsAdapter(true)
                 val layoutManager = GridLayoutManager(this, SPAN_COUNT)
-                binding.imageRecycler.layoutManager = layoutManager
+                Pair(adapter, layoutManager)
             } else {
-                val adapter = SharedItemsListAdapter()
-                adapter.items = it.items
-                adapter.authHeader = it.authHeader
-                binding.imageRecycler.adapter = adapter
-
+                val adapter = SharedItemsAdapter(false)
                 val layoutManager = LinearLayoutManager(this)
                 layoutManager.orientation = LinearLayoutManager.VERTICAL
-                binding.imageRecycler.layoutManager = layoutManager
+                Pair(adapter, layoutManager)
             }
+
+            pair.first.items = it.items
+            pair.first.authHeader = it.authHeader
+            binding.imageRecycler.adapter = pair.first
+            binding.imageRecycler.layoutManager = pair.second
```

Comment:
You can avoid the Pair entirely:

```kotlin
            val showGrid = viewModel.currentItemType == SharedItemType.MEDIA
            val layoutManager = if (showGrid) {
                GridLayoutManager(this, SPAN_COUNT)
            } else {
                LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)
            }

            val adapter = SharedItemsAdapter(showGrid).apply {
                items = it.items
                authHeader = it.authHeader
            }
            binding.imageRecycler.adapter = adapter
            binding.imageRecycler.layoutManager = layoutManager
```
I'm not entirely sure if the decision between grid and list should be done in the ViewModel or in the Activity, but let's leave it here for now.

Reply:
That's much better. Thanks. 

In my view this part is all about bindings. So it should not be done in the ViewModel. Maybe this can also be moved into the adapter?

Reply:
> In my view this part is all about bindings. So it should not be done in the ViewModel. Maybe this can also be moved into the adapter?

It's currently done in the `Activity`, and I think it probably should remain there, because the `layoutManager` is set to the `recyclerView`, which only the `Activity` should have access to.

You may want to move it to a new method to clear `onCreate` a bit, if you want to tidy up.